src/Native/Unix/System.Native/pal_random.c

```diff
@@ -2,109 +2,206 @@
+#include <stdlib.h>
+#include <string.h>
 #include <sys/stat.h>
-#include <fcntl.h>
-#include <assert.h>
-#include <unistd.h>
 #include <time.h>
-#include <errno.h>
+#include <unistd.h>
 
 #include "pal_config.h"
 #include "pal_random.h"
+
+#if HAVE_SYS_GETRANDOM
+#include <sys/syscall.h>
+
+#ifndef GRND_NONBLOCK
+#define GRND_NONBLOCK 0x1
+#endif // GRND_NONBLOCK
+#endif // HAVE_SYS_GETRANDOM
+
 /*
 
 Generate random bytes. The generated bytes are not cryptographically strong.
 
 */
 
-void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t* buffer, int32_t bufferLength)
+#if HAVE_ARC4RANDOM_BUF
+static void FillWithArc4RandomBuf(uint8_t* buffer, int32_t bufferLength)
 {
     assert(buffer != NULL);
 
-#if HAVE_ARC4RANDOM_BUF
     arc4random_buf(buffer, (size_t)bufferLength);
+}
 #else
-    static volatile int rand_des = -1;
-    long num = 0;
-    static bool sMissingDevURandom;
-    static bool sInitializedMRand;
+static int GetUrandomFileDescriptor(void)
+{
+    static int fd = -1;
 
-    if (!sMissingDevURandom)
+    if (fd < 0)
     {
-        if (rand_des == -1)
+        do
         {
-            int fd;
-
-            do
-            {
 #if HAVE_O_CLOEXEC
-                fd = open("/dev/urandom", O_RDONLY, O_CLOEXEC);
+            fd = open("/dev/urandom", O_RDONLY | O_CLOEXEC);
 #else
-                fd = open("/dev/urandom", O_RDONLY);
-                fcntl(fd, F_SETFD, FD_CLOEXEC);
+            fd = open("/dev/urandom", O_RDONLY);
+            fcntl(fd, F_SETFD, FD_CLOEXEC);
 #endif
-            }
-            while ((fd == -1) && (errno == EINTR));
-
-            if (fd != -1)
-            {
-                int expected = -1;
-                if (!__atomic_compare_exchange_n(&rand_des, &expected, fd, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
-                {
-                    // Another thread has already set the rand_des
-                    close(fd);
-                }
-            }
-            else if (errno == ENOENT)
-            {                
-                sMissingDevURandom = true;
-            }
         }
+        while ((fd == -1) && (errno == EINTR));
+    }
+
+    return fd;
+}
+
+#if HAVE_SYS_GETRANDOM
+static bool IsGetRandomAvailable(void)
+{
+    uint8_t buffer;
+    long r = syscall(SYS_getrandom, &buffer, 1, GRND_NONBLOCK);
+
+    if (r == 1)
+    {
+        return true;
+    }
+    if ((r == -1) && (errno != ENOSYS))
+    {
+        // At this point, we're only interested if getrandom(2) is available
+        // in this kernel.  If errno isn't ENOSYS, then we're good.
+        return true;
+    }
+
+    return false;
+}
 
-        if (rand_des != -1)
+static bool TryFillWithGetRandom(uint8_t* buffer, int32_t bufferLength)
+{
+    size_t length = (size_t)bufferLength;
+
+    do
+    {
+        long n = syscall(SYS_getrandom, buffer, length, GRND_NONBLOCK);
+
+        if (n == -1)
         {
-            int32_t offset = 0;
-            do
+            if (errno == EINTR)
```

Comment:
According to the docs, `getrandom` will return -1 and `errno` will be `EAGAIN` when `GRND_NONBLOCK` is used and the entropy pool is not available. Should this include `EAGAIN` to try again?

Reply:
Yes, it's often the case during early boot scenarios, or on embedded systems. However, calling getrandom() in a tight loop if it keeps returning EAGAIN might turn into an infinite loop in such cases; I'm not sure it's advisable.

Reply:
I have at least one VM that can take minutes to fully initialize.

Ubuntu 16.04.6
```
$ dmesg | grep random
[    1.844973] random: udevadm: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.846829] random: udevadm: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.849528] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.850996] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.852320] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.864731] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.866058] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.867459] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.870948] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[    1.872280] random: systemd-udevd: uninitialized urandom read (16 bytes read, 2 bits of entropy available)
[  180.310815] random: nonblocking pool is initialized
```

Newer kernels trust the CPU, so it's less likely:

Ubuntu 19.04:
```
$ dmesg | grep random
[    0.017218] random: get_random_bytes called from start_kernel+0x97/0x516 with crng_init=0
[    0.536646] random: crng done (trusting CPU's manufacturer)
```

Though not necessarily:

Ubuntu 18.04:

```
$ dmesg | grep random
[    0.000000] random: get_random_bytes called from start_kernel+0x99/0x4fd with crng_init=0
[    1.452697] random: fast init done
[    1.454713] random: systemd-udevd: uninitialized urandom read (16 bytes read)
[    1.462508] random: systemd-udevd: uninitialized urandom read (16 bytes read)
[    1.466497] random: systemd-udevd: uninitialized urandom read (16 bytes read)
[   10.935718] random: crng init done
[   10.935721] random: 7 urandom warning(s) missed due to ratelimiting
```

So the EAGAIN either needs to fail over (e.g. to "dirty reads" from /dev/urandom), or the read should just be a blocking read.  But it's definitely an important case to handle.

Reply:
(Possibly it's already being handled OK, just that the assert needs to be removed)

Reply:
The assert was just a copy from the read(/dev/urandom) codepath.  Maybe a good idea to remove it from there.

BTW, not sure if you're able to set this up in your VM, but there is a virtio-rng device that some hypervisors implement (qemu-kvm being one of them) that provide entropy from the host to the guest; the Linux kernel is able to use it if detected.

Reply:
> How does that jive with it potentially taking minutes to initialize? I don't see how that's viable.

Only at boot.  Once a system has initialized it'll never block again for any process, until shutdown.

/dev/urandom will give "dirty random" (which we currently shrug and hope is adequate, and it probably is), getrandom() will (IIUC) leave the data uninitialized (aka simply not write to it).  My comment is just that we can't leave it uninitialized, so either loop, block, or read /dev/urandom.  Looping and blocking are effectively the same thing.

Reply:
> so either loop, block, or read /dev/urandom

And my comment is we shouldn't block or loop if we may do so for any measurable period of time.

Reply:
@bartonjs Is the "no entropy available" message you're receiving on a fresh VM, or on a VM that has been rebooted? Per http://man7.org/linux/man-pages/man4/random.4.html:

>  If a seed file is saved across reboots as recommended below (all
       major Linux distributions have done this since 2000 at least), the
       output is cryptographically secure against attackers without local
       root access as soon as it is reloaded in the boot sequence, and
       perfectly adequate for network encryption session keys.

IMO blocking is the preferred behavior, especially if it happens only very rarely. If somebody's calling our APIs expecting high-entropy random values then we need to give them high-entropy random values. We can't pull a sleight of hand and give them low-entropy values.

Reply:
> IMO blocking is the preferred behavior

This caused significant problems when we were initially bringing up .NET on Linux.  We initially had a blocking implementation, and start-up of the process could end up getting blocked for seconds, until the user wildly moved their mouse around to introduce entropy into the system.

Maybe what we're talking about here is a special case somehow, or maybe things have been improved on all distros/versions we now support, but I don't believe blocking waiting for sufficient entropy is viable.

Reply:
FWIW, the Python crew considered this same problem a few years ago. They took a breaking change between Python 3.5.2 and Python 3.6 in order to make the `os.urandom()` function block on low-entropy scenarios because of the security risks inherent in the original behavior. See https://www.python.org/dev/peps/pep-0524/ for more information.

Reply:
`os.urandom` is returns bytes suitable for cryptographic use (https://docs.python.org/2/library/os.html#os.urandom).

`SystemNative_GetNonCryptographicallySecureRandomBytes` is not meant to be suitable for cryptographic use as the name suggests.

Reply:
If this API isn't intended to produce strong randomness, we should reconsider having `Guid.NewGuid` depend on it. We know from experience that for better or worse, devs use `Guid.NewGuid` to produce random identifiers and expect those identifiers to have strong randomness.